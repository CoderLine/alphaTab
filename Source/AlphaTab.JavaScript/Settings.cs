using AlphaTab.Collections;
using AlphaTab.Haxe.Js;
using AlphaTab.Util;
using Phase;

namespace AlphaTab
{
    public partial class CoreSettings
    {
        /// <summary>
        /// Gets or sets the script file url that will be used to spawn the workers.
        /// </summary>
        [JsonName("scriptFile")]
        public string ScriptFile { get; set; }

        /// <summary>
        /// Gets or sets the url to the fonts that will be used to generate the alphaTab font style.
        /// </summary>
        [JsonName("fontDirectory")]
        public string FontDirectory { get; set; }

        /// <summary>
        /// Gets or sets whether lazy loading for displayed elements is enabled.
        /// </summary>
        [JsonName("enableLazyLoading")]
        public bool EnableLazyLoading { get; set; } = true;

        public CoreSettings()
        {
            var global = Lib.Global;
            if (global.document && global.ALPHATAB_ROOT)
            {
                ScriptFile = global.ALPHATAB_ROOT;
                ScriptFile = EnsureFullUrl(ScriptFile);
                ScriptFile = AppendScriptName(ScriptFile);
            }
            else
            {
                ScriptFile = Environment.ScriptFile;
            }

            if (global.document && global.ALPHATAB_FONT)
            {
                FontDirectory = global.ALPHATAB_FONT;
                FontDirectory = EnsureFullUrl(FontDirectory);
            }
            else
            {
                FontDirectory = ScriptFile;
                if (!string.IsNullOrEmpty(FontDirectory))
                {
                    var lastSlash = FontDirectory.LastIndexOf('/');
                    if (lastSlash >= 0)
                    {
                        FontDirectory = FontDirectory.Substring(0, lastSlash) + "/Font/";
                    }
                }
            }
        }

        private static string EnsureFullUrl(string relativeUrl)
        {
            var global = Lib.Global;
            if (!relativeUrl.StartsWith("http") && !relativeUrl.StartsWith("https") && !relativeUrl.StartsWith("file"))
            {
                var root = new StringBuilder();
                root.Append(global.location.protocol);
                root.Append("//");
                if (global.location.hostname)
                {
                    root.Append(global.location.hostname);
                }

                if (global.location.port)
                {
                    root.Append(":");
                    root.Append(global.location.port);
                }

                // as it is not clearly defined how slashes are treated in the location object
                // better be safe than sorry here
                string directory = global.location.pathname.split("/").slice(0, -1).join("/");
                if (directory.Length > 0)
                {
                    if (!directory.StartsWith("/"))
                    {
                        root.Append("/");
                    }

                    root.Append(directory);
                }


                if (!relativeUrl.StartsWith("/"))
                {
                    root.Append("/");
                }

                root.Append(relativeUrl);

                return root.ToString();
            }

            return relativeUrl;
        }

        private static string AppendScriptName(string url)
        {
            // append script name
            if (!string.IsNullOrEmpty(url) && !url.EndsWith(".js"))
            {
                if (!url.EndsWith("/"))
                {
                    url += "/";
                }

                url += "AlphaTab.js";
            }

            return url;
        }

    }

    public partial class PlayerSettings
    {
        /// <summary>
        /// Gets or sets the URl of the sound font to be loaded.
        /// </summary>
        [JsonName("soundFont")]
        public string SoundFont { get; set; }

        /// <summary>
        /// Gets or sets the element that should be used for scrolling.
        /// </summary>
        [JsonName("scrollElement")]
        public string ScrollElement { get; set; } = "html,body";
    }

    /// <summary>
    /// This public class contains instance specific settings for alphaTab
    /// </summary>
    public partial class Settings
    {
        public static Settings FromJson(dynamic json)
        {
            return null; // dynamically implemented via macro
        }

        public void FillFromJson(dynamic json)
        {
            // dynamically implemented via macro
        }

        public static object ToJson(Settings settings)
        {
            return null; // dynamically implemented via macro
        }

        public void FillFromDataAttributes(dynamic dataAttributes)
        {
            object dataAttributesObject = dataAttributes;
            var keys = Platform.Platform.JsonKeys(dataAttributesObject);
            foreach (var key in keys)
            {
                SetProperty(key.ToLower(), dataAttributes[key]);
            }
        }

        public bool SetProperty(string property, dynamic value)
        {
            // dynamically implemented via macro
            return false;
        }

    }
}
