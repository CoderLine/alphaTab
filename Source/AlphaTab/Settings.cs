using System.Runtime.CompilerServices;
using AlphaTab.Collections;

namespace AlphaTab
{
    /// <summary>
    /// This public class contains instance specific settings for alphaTab
    /// </summary>
    public class Settings
    {
        /// <summary>
        /// Sets the zoom level of the rendered notation
        /// </summary>
        [IntrinsicProperty]
        public float Scale { get; set; }

        /// <summary>
        /// The initial size of the canvas during loading or the width when {@see autoSize} set to false
        /// </summary>
        [IntrinsicProperty]
        public int Width { get; set; }

        /// <summary>
        /// The initial size of the canvas during loading or the fixed height on some layouts. 
        /// </summary>
        [IntrinsicProperty]
        public int Height { get; set; }

        /// <summary>
        /// The engine which should be used to render the the tablature. 
        /// <ul>
        ///  <li><strong>default</strong> - Platform specific default engine</li>
        ///  <li><strong>html5</strong> - Canvas with VRML Fallback</li>
        ///  <li><strong>svg</strong> -  SVG </li>
        /// </ul>
        /// </summary>
        [IntrinsicProperty]
        public string Engine { get; set; }

        /// <summary>
        /// The layout specific settings
        /// </summary>
        [IntrinsicProperty]
        public LayoutSettings Layout { get; set; }

        /// <summary>
        /// The staves to create for each row.
        /// </summary>
        [IntrinsicProperty]
        public FastList<StaveSettings> Staves { get; set; }

        public static Settings Defaults
        {
            get
            {
                var settings = new Settings();

                settings.Scale = 1.0f;
                settings.Width = 600;
                settings.Height = 200;
                settings.Engine = "default";

                settings.Layout = LayoutSettings.Defaults;

                settings.Staves = new FastList<StaveSettings>();

                settings.Staves.Add(new StaveSettings("marker"));
                //settings.staves.Add(new StaveSettings("triplet-feel"));
                settings.Staves.Add(new StaveSettings("tempo"));
                settings.Staves.Add(new StaveSettings("text"));
                settings.Staves.Add(new StaveSettings("chords"));
                settings.Staves.Add(new StaveSettings("trill"));
                settings.Staves.Add(new StaveSettings("beat-vibrato"));
                settings.Staves.Add(new StaveSettings("note-vibrato"));
                settings.Staves.Add(new StaveSettings("alternate-endings"));

                settings.Staves.Add(new StaveSettings("score"));

                settings.Staves.Add(new StaveSettings("crescendo"));
                settings.Staves.Add(new StaveSettings("dynamics"));
                settings.Staves.Add(new StaveSettings("trill"));
                settings.Staves.Add(new StaveSettings("beat-vibrato"));
                settings.Staves.Add(new StaveSettings("note-vibrato"));
                settings.Staves.Add(new StaveSettings("tap"));
                settings.Staves.Add(new StaveSettings("fade-in"));
                settings.Staves.Add(new StaveSettings("let-ring"));
                settings.Staves.Add(new StaveSettings("palm-mute"));

                settings.Staves.Add(new StaveSettings("tab"));

                settings.Staves.Add(new StaveSettings("pick-stroke"));
                //settings.staves.Add(new StaveSettings("fingering"));

                return settings;
            }
        }

#if JavaScript
        [System.Runtime.CompilerServices.InlineCodeAttribute("{property} in {json}")]
        private static bool JsonExists(dynamic json, string property)
        {
            return false;
        }

        [InlineCode("Object.keys({json})")]
        private static string[] JsonKeys(dynamic json)
        {
            return null;
        }

        public static Settings fromJson(dynamic json)
        {
            if (json is Settings)
            {
                return (Settings)json;
            }

            var settings = Defaults;

            if (!json) return settings;
            if (JsonExists(json, "scale")) settings.Scale = json.scale;
            if (JsonExists(json, "width")) settings.Width = json.width;
            if (JsonExists(json, "height")) settings.Height = json.height;
            if (JsonExists(json, "engine")) settings.Engine = json.engine;

            if (JsonExists(json, "layout"))
            {
                if (json.layout is string)
                {
                    settings.Layout.Mode = json.layout;
                }
                else
                {
                    if (json.layout.mode) settings.Layout.Mode = json.layout.mode;
                    if (json.layout.additionalSettings)
                    {
                        string[] keys = JsonKeys(json.layout.additionalSettings);
                        foreach (var key in keys)
                        {
                            settings.Layout.AdditionalSettings[key] = json.layout.additionalSettings[key];
                        }
                    }
                }
            }

            if (JsonExists(json, "staves"))
            {
                settings.Staves = new FastList<StaveSettings>();
                string[] keys = JsonKeys(json.staves);
                foreach (var key in keys)
                {
                    var val = json.staves[key];
                    if (val is string)
                    {
                        settings.Staves.Add(new StaveSettings(val));
                    }
                    else
                    {
                        if (val.id)
                        {
                            var staveSettings = new StaveSettings(val.id);
                            if (val.additionalSettings)
                            {
                                string[] keys2 = JsonKeys(val.additionalSettings);
                                foreach (var key2 in keys2)
                                {
                                    staveSettings.AdditionalSettings[key2] = val.additionalSettings[key2];
                                }
                            }
                        }
                    }
                }
            }

            return settings;
        }
#endif

    }

    public class LayoutSettings
    {
        /// <summary>
        /// The layouting mode used to arrange the the notation.
        /// <ul>
        ///  <li><strong>page</strong> - Bars are aligned in rows using a fixed width</li>
        ///  <li><strong>horizontal</strong> - Bars are aligned horizontally in one row</li>
        /// </ul>
        /// </summary>
        [IntrinsicProperty]
        public string Mode { get; set; }

        /// <summary>
        /// Additional layout mode specific settings.
        /// <strong>mode=page</strong>
        /// <ul>
        ///  <li><strong>autoSize</strong> - Whether the width of the canvas should be automatically determined by the used layout. (bool, default:true)</li>
        ///  <li><strong>barsPerRow</strong> - Limit the displayed bars per row, <em>-1 for sized based limit<em> (integer, default:-1)</li>
        ///  <li><strong>spacingScale</strong> - Control how closely glyphs are spaced (float, default:1.0f)</li>
        ///  <li><strong>start</strong> - The bar start index to start layouting with (integer: default: 0)</li>
        ///  <li><strong>count</strong> - The amount of bars to render overall, <em>-1 for all till the end</em>  (integer, default:-1)</li>
        ///  <li><strong>hideInfo</strong> - Render the song information or not (boolean, default:true)</li>
        /// </ul>
        /// <strong>mode=horizontal</strong>
        /// <ul>
        ///  <li><strong>start</strong> - The bar start index to start layouting with (integer: default: 0)</li>
        ///  <li><strong>count</strong> - The amount of bars to render overall, <em>-1 for all till the end</em>  (integer, default:-1)</li>
        /// </ul>
        /// </summary>
        [IntrinsicProperty]
        public FastDictionary<string, object> AdditionalSettings { get; set; }

        [IncludeGenericArguments(false)]
        public T Get<T>(string key, T def)
        {
            if (AdditionalSettings.ContainsKey(key))
            {
                return (T)(AdditionalSettings[key]);
            }
            return def;
        }

        public LayoutSettings()
        {
            AdditionalSettings = new FastDictionary<string, object>();
        }

        public static LayoutSettings Defaults
        {
            get
            {
                var settings = new LayoutSettings();
                settings.Mode = "page";
                return settings;
            }
        }
    }

    public class StaveSettings
    {
        /// <summary>
        /// The stave identifier.
        /// Default Staves: 
        /// <ul>
        ///  <li><strong>marker</strong> - Renders section markers</li>
        ///  <li><strong>triplet-feel</strong> - Renders triplet feel indicators</li>
        ///  <li><strong>tempo</strong> - Renders a tempo identifier</li>
        ///  <li><strong>text</strong> - Renders custom text annotations</li>
        ///  <li><strong>chords</strong> - Renders chord names</li>
        ///  <li><strong>beat-vibrato</strong> - Renders beat vibrato symbols</li>
        ///  <li><strong>note-vibrato</strong> - Renders note vibrato symbols</li>
        ///  <li><strong>tuplet</strong> - Renders tuplet indicators</li>
        ///  <li><strong>score</strong> - Renders default music notation</li>
        ///  <li><strong>crescendo</strong> - Renders crescendo and decresencod</li>
        ///  <li><strong>dynamics</strong> - Renders dynamics markers</li>
        ///  <li><strong>tap</strong> - Renders tap/slap/pop indicators</li>
        ///  <li><strong>fade-in</strong> - Renders fade-in indicators</li>
        ///  <li><strong>let-ring</strong> - Renders let-ring indicators</li>
        ///  <li><strong>palm-mute</strong> - Renders palm-mute indicators</li>
        ///  <li><strong>tab</strong> - Renders guitar tablature</li>
        ///  <li><strong>fingering</strong> - Renders finger indicators</li>
        /// </ul>
        /// </summary>
        [IntrinsicProperty]
        public string Id { get; set; }

        /// <summary>
        /// Additional stave specific settings
        /// <strong>id=tab</strong>
        /// <ul>
        ///  <li><strong>rhythm</strong> - Renders rhythm beams to tablature notes</li>
        /// </ul>
        /// </summary>
        [IntrinsicProperty]
        public FastDictionary<string, object> AdditionalSettings { get; set; }

        public StaveSettings(string id)
        {
            Id = id;
            AdditionalSettings = new FastDictionary<string, object>();
        }
    }
}
