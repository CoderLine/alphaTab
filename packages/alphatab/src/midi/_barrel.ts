export { BeatTickLookup, BeatTickLookupItem } from '@src/midi/BeatTickLookup';
export { MasterBarTickLookup, MasterBarTickLookupTempoChange } from '@src/midi/MasterBarTickLookup';
export {
    MidiTickLookup,
    MidiTickLookupFindBeatResult,
    MidiTickLookupFindBeatResultCursorMode
} from '@src/midi/MidiTickLookup';
export { MidiFile, MidiFileFormat, MidiTrack } from '@src/midi/MidiFile';
export { ControllerType } from '@src/midi/ControllerType';
export {
    MidiEvent,
    MidiEventType,
    TimeSignatureEvent,
    AlphaTabRestEvent,
    AlphaTabMetronomeEvent,
    NoteEvent,
    NoteOnEvent,
    NoteOffEvent,
    ControlChangeEvent,
    ProgramChangeEvent,
    TempoChangeEvent,
    PitchBendEvent,
    NoteBendEvent,
    EndOfTrackEvent,
    AlphaTabSysExEvent
} from '@src/midi/MidiEvent';
export {
    DeprecatedMidiEvent,
    MetaEventType,
    MetaEvent,
    MetaDataEvent,
    MetaNumberEvent,
    Midi20PerNotePitchBendEvent,
    SystemCommonType,
    SystemCommonEvent,
    AlphaTabSystemExclusiveEvents,
    SystemExclusiveEvent
} from '@src/midi/DeprecatedEvents';
export { MidiFileGenerator } from '@src/midi/MidiFileGenerator';
export { AlphaSynthMidiFileHandler } from '@src/midi/AlphaSynthMidiFileHandler';
export type { IMidiFileHandler } from '@src/midi/IMidiFileHandler';
