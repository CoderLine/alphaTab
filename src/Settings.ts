import { CoreSettings } from '@src/CoreSettings';
import { DisplaySettings } from '@src/DisplaySettings';
import { ImporterSettings } from '@src/ImporterSettings';
import { FingeringMode, NotationMode, NotationSettings, NotationElement } from '@src/NotationSettings';
import { PlayerSettings } from '@src/PlayerSettings';

/**
 * This public class contains instance specific settings for alphaTab
 * @json
 */
export class Settings {
    /**
     * The core settings control the general behavior of alphatab like
     * what modules are active.
     * @json_on_parent
     * @json_partial_names
     */
    public readonly core: CoreSettings = new CoreSettings();

    /**
     * The display settings control how the general layout and display of alphaTab is done.
     * @json_on_parent
     * @json_partial_names
     */
    public readonly display: DisplaySettings = new DisplaySettings();

    /**
     * The notation settings control how various music notation elements are shown and behaving.
     * @json_partial_names
     */
    public readonly notation: NotationSettings = new NotationSettings();

    /**
     * All settings related to importers that decode file formats.
     * @json_partial_names
     */
    public readonly importer: ImporterSettings = new ImporterSettings();

    /**
     * Contains all player related settings
     * @json_partial_names
     */
    public player: PlayerSettings = new PlayerSettings();

    public setSongBookModeSettings(): void {
        this.notation.notationMode = NotationMode.SongBook;
        this.notation.smallGraceTabNotes = false;
        this.notation.fingeringMode = FingeringMode.SingleNoteEffectBand;
        this.notation.extendBendArrowsOnTiedNotes = false;
        this.notation.elements.set(NotationElement.ParenthesisOnTiedBends, false);
        this.notation.elements.set(NotationElement.TabNotesOnTiedBends, false);
        this.notation.elements.set(NotationElement.ZerosOnDiveWhammys, true);
    }

    public static get songBook(): Settings {
        let settings: Settings = new Settings();
        settings.setSongBookModeSettings();
        return settings;
    }
}
