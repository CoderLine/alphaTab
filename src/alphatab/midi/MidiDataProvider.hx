/*
 * This file is part of alphaTab.
 *
 *  alphaTab is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  alphaTab is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with alphaTab.  If not, see <http://www.gnu.org/licenses/>.
 */
package alphatab.midi;
import alphatab.model.Song;
import alphatab.model.SongFactory;
import alphatab.tablature.model.MeasureClickable;

/**
 * This wrapper allows easy converting of a song into a midi message list used by the player.
 */
class MidiDataProvider 
{
	public static function getSongMidiData(song:Song, factory:SongFactory, measureMap:Array<MeasureClickable>):String
	{
		var parser:MidiSequenceParser = new MidiSequenceParser(factory, song, MidiSequenceParserFlags.DEFAULT_PLAY_FLAGS,
											   100, 0);
		var sequence:MidiSequenceHandler = new MidiSequenceHandler(song.tracks.length + 2);
		parser.parse(sequence, measureMap);
		return sequence.commands;
	}
}