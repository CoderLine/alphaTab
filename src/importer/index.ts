import { ScoreImporter } from '@src/importer/ScoreImporter';
import { ScoreLoader } from '@src/importer/ScoreLoader';
import { UnsupportedFormatError } from '@src/importer/UnsupportedFormatError';

export { ScoreImporter, ScoreLoader, UnsupportedFormatError };
