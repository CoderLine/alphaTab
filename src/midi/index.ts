export { BeatTickLookup } from '@src/midi/BeatTickLookup';
export { MasterBarTickLookup } from '@src/midi/MasterBarTickLookup';
export { MidiTickLookup, MidiTickLookupFindBeatResult } from '@src/midi/MidiTickLookup';
export { MidiFile } from '@src/midi/MidiFile';
export { ControllerType } from '@src/midi/ControllerType';
export { MetaDataEvent } from '@src/midi/MetaDataEvent';
export { MetaEvent, MetaEventType } from '@src/midi/MetaEvent';
export { MetaNumberEvent } from '@src/midi/MetaNumberEvent';
export { MidiEvent, MidiEventType } from '@src/midi/MidiEvent';
export { Midi20PerNotePitchBendEvent } from '@src/midi/Midi20PerNotePitchBendEvent';
export { SystemCommonEvent, SystemCommonType } from '@src/midi/SystemCommonEvent';
export { SystemExclusiveEvent } from '@src/midi/SystemExclusiveEvent';
export { MidiFileGenerator } from '@src/midi/MidiFileGenerator';
export { AlphaSynthMidiFileHandler } from '@src/midi/AlphaSynthMidiFileHandler';
