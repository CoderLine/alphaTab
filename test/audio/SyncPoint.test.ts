import { AlphaTexImporter } from '@src/importer/AlphaTexImporter';
import { MidiFileGenerator } from '@src/midi/MidiFileGenerator';
import { Settings } from '@src/Settings';
import { FlatMidiEventGenerator } from '@test/audio/FlatMidiEventGenerator';
import { expect } from 'chai';

describe('SyncPointTests', () => {
    it('sync-point-update', () => {
        // MidiFileSequencer
        // sync points and tempo changes -> expect interpolation
    });

    it('backing-track-time-mapping', () => {
        // MidiFileSequencer
        // do a variety of lookups along the time axis.
        // - sequentially (playback)
        // - jumps (seeks back and forth)
        // check
        // - updated syncPointIndex
        // - interpolated time
        // - reverse lookup with mainTimePositionToBackingTrack
    });

    function syncPointTestScore() {
        const settings = new Settings();
        const importer = new AlphaTexImporter();

        // Tempo 120:  2000ms per bar
        // Tempo 60:   4000ms per bar
        // Tempo 80:   3000ms per bar

        importer.initFromString(
            `
            .
            \\tempo 120

            C4 * 4 |
            C4 * 4 |
            C4 * 4 |
            C4 * 4 |

            \\tempo 60
            C4 * 4 |
            C4 * 4 |
            C4 * 4 |
            C4 * 4 |

            \\tempo 80
            \\ro 
            C4 * 4 |
            C4 * 4 |
            C4 * 4 |
            \\rc 2
            C4 * 4 |

            .

            \\sync 0  0 0     
            \\sync 2  0 8000  // 00:08 instead of 00:04 -> should give half BPM of 60 
                              
            \\sync 4  0 10000  // restore tempo 60 at 00:10  
            \\sync 5  0 14000  // 00:14 

            \\sync 8  0 26000 // 00:26 on first occurence speed up to 120bpm 
            \\sync 9  0 28000 // 00:28 
            
            \\sync 9  1 36000 // 00:36 on second occurence of bar 9 restore tempo 80 until end
            \\sync 10 1 39000 // 00:39
            
        `,
            settings
        );
        const score = importer.readScore();

        return score;
    }

    it('sync-point-generation', () => {
        const score = syncPointTestScore();

        const handler = new FlatMidiEventGenerator();
        const generator = new MidiFileGenerator(score, new Settings(), handler);
        generator.generate();

        expect(generator.syncPoints).toMatchSnapshot();

        const update = MidiFileGenerator.generateSyncPoints(score);

        expect(generator.syncPoints.length).to.equal(update.length);
        for (let i = 0; i < generator.syncPoints.length; i++) {
            expect(update[i].masterBarIndex).to.equal(generator.syncPoints[i].masterBarIndex);
            expect(update[i].masterBarOccurence).to.equal(generator.syncPoints[i].masterBarOccurence);
            expect(update[i].syncBpm).to.equal(generator.syncPoints[i].syncBpm);
            expect(update[i].syncTime).to.equal(generator.syncPoints[i].syncTime);
            expect(update[i].synthBpm).to.equal(generator.syncPoints[i].synthBpm);
            expect(update[i].synthTick).to.equal(generator.syncPoints[i].synthTick);
            expect(update[i].synthTime).to.equal(generator.syncPoints[i].synthTime);
        }
    });

    it('modified-tempo-lookup', () => {
        const score = syncPointTestScore();
        expect(MidiFileGenerator.buildModifiedTempoLookup(score)).toMatchSnapshot();
    });

    it('playback-normal-backing-track', () => {
        // BackingTrackPlayer
        // play a variety of songs artificially
        // check
        // - the positionChanged event list provided
    });

    it('playback-slow-backing-track', () => {
        // BackingTrackPlayer
        // play a variety of songs artificially (other 0.5 speed)
        // check
        // - the positionChanged event list provided
    });

    it('seek-normal-backing-track', () => {
        // BackingTrackPlayer
        // seek to a given position
        // check
        // - the positionChanged event list provided
    });

    it('seek-slow-backing-track', () => {
        // BackingTrackPlayer
        // seek to a given position (playbackSpeed 0.5)
        // check
        // - the positionChanged event list provided
    });

    it('playback-normal-external-media', () => {
        // ExternalMediaPlayer
        // play a variety of songs artificially
        // check
        // - the positionChanged event list provided
    });

    it('playback-slow-external-media', () => {
        // ExternalMediaPlayer
        // play a variety of songs artificially (other 0.5 speed)
        // check
        // - the positionChanged event list provided
    });

    it('seek-normal-external-media', () => {
        // ExternalMediaPlayer
        // seek to a given position
        // check
        // - the positionChanged event list provided
    });

    it('seek-slow-external-media', () => {
        // ExternalMediaPlayer
        // seek to a given position (playbackSpeed 0.5)
        // check
        // - the positionChanged event list provided
    });
});
