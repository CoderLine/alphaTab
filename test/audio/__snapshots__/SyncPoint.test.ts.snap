// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncPointTests modified-tempo-lookup 1`] = `
Map {
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 0,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 0,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 0,
    "synthBpm": 120,
    "synthTick": 0,
    "synthTime": 0,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 8000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 2,
    "masterBarOccurence": 0,
    "syncBpm": 240,
    "syncTime": 8000,
    "synthBpm": 120,
    "synthTick": 7680,
    "synthTime": 4000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 10000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 4,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 10000,
    "synthBpm": 60,
    "synthTick": 15360,
    "synthTime": 8000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 14000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 5,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 14000,
    "synthBpm": 60,
    "synthTick": 19200,
    "synthTime": 12000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 26000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 8,
    "masterBarOccurence": 0,
    "syncBpm": 120,
    "syncTime": 26000,
    "synthBpm": 80,
    "synthTick": 30720,
    "synthTime": 24000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 0,
      "millisecondOffset": 28000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 9,
    "masterBarOccurence": 0,
    "syncBpm": 120,
    "syncTime": 28000,
    "synthBpm": 80,
    "synthTick": 34560,
    "synthTime": 27000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 1,
      "millisecondOffset": 36000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 9,
    "masterBarOccurence": 1,
    "syncBpm": 80,
    "syncTime": 36000,
    "synthBpm": 80,
    "synthTick": 49920,
    "synthTime": 39000,
  },
  Automation {
    "isLinear": false,
    "ratioPosition": 0,
    "syncPointValue": SyncPointData {
      "barOccurence": 1,
      "millisecondOffset": 39000,
    },
    "text": "",
    "type": 4,
    "value": 0,
  } => BackingTrackSyncPoint {
    "masterBarIndex": 10,
    "masterBarOccurence": 1,
    "syncBpm": 80,
    "syncTime": 39000,
    "synthBpm": 80,
    "synthTick": 53760,
    "synthTime": 42000,
  },
}
`;

exports[`SyncPointTests sync-point-generation 1`] = `
Array [
  BackingTrackSyncPoint {
    "masterBarIndex": 0,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 0,
    "synthBpm": 120,
    "synthTick": 0,
    "synthTime": 0,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 2,
    "masterBarOccurence": 0,
    "syncBpm": 240,
    "syncTime": 8000,
    "synthBpm": 120,
    "synthTick": 7680,
    "synthTime": 4000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 4,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 10000,
    "synthBpm": 60,
    "synthTick": 15360,
    "synthTime": 8000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 5,
    "masterBarOccurence": 0,
    "syncBpm": 60,
    "syncTime": 14000,
    "synthBpm": 60,
    "synthTick": 19200,
    "synthTime": 12000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 8,
    "masterBarOccurence": 0,
    "syncBpm": 120,
    "syncTime": 26000,
    "synthBpm": 80,
    "synthTick": 30720,
    "synthTime": 24000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 9,
    "masterBarOccurence": 0,
    "syncBpm": 120,
    "syncTime": 28000,
    "synthBpm": 80,
    "synthTick": 34560,
    "synthTime": 27000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 9,
    "masterBarOccurence": 1,
    "syncBpm": 80,
    "syncTime": 36000,
    "synthBpm": 80,
    "synthTick": 49920,
    "synthTime": 39000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 10,
    "masterBarOccurence": 1,
    "syncBpm": 80,
    "syncTime": 39000,
    "synthBpm": 80,
    "synthTick": 53760,
    "synthTime": 42000,
  },
  BackingTrackSyncPoint {
    "masterBarIndex": 11,
    "masterBarOccurence": 0,
    "syncBpm": 80,
    "syncTime": 45000,
    "synthBpm": 80,
    "synthTick": 61440,
    "synthTime": 48000,
  },
]
`;
